### ISSUE TITLE
Switch test utils no longer working from v0.15.1

### ISSUE THREAD

**Issue by mhemeryck** [django-waffle/django-waffle#330]
**tl; dr** Test utility `override_switch` does not switch off given switch.

# Background

I have a setup where I wanted a simple switch to toggle certain behavior. In my test setup, I use the standard `override_switch` test utility to check the behavior under different conditions of the switch.

# How to replicate
Let's say you have these tests:

```python
from pytest import mark
from waffle import switch_is_active
from waffle.testutils import override_switch
    
    
@mark.django_db
def test_feature_flag_override_true():
    with override_switch("myswitch", active=True):
        assert switch_is_active("myswitch")
    

@mark.django_db
def test_feature_flag_override_false():
    with override_switch("myswitch", active=False):
        assert not switch_is_active("myswitch")
```
# Expected
Expected behavior would be they both pass.

# Actual
With v0.14, tests both pass. For v0.15.1:
  - `test_feature_flag_override_true`: passes
  - `test_feature_flag_override_false`: does _not_ pass, i.e. the override trying to disable the switch does not disable it (stays `True`).

```
______________________________________________________________________________ test_feature_flag_override_false ______________________________________________________________________________

    @mark.django_db
    def test_feature_flag_override_false():
        with override_switch("myswitch", active=False):
>           assert not switch_is_active("myswitch")
E           AssertionError: assert not True
E            +  where True = switch_is_active('myswitch')
```

Can you confirm the issue? For now, I shall be using v0.14 in my own project.

# Context

Some more context:
* `Django==1.11.15`
* `pytest==4.2.0`
* `pytest-django==3.4.7`
* Python 2.7.15rc1 (I know :see_no_evil: )

**Comment by mhemeryck** [comment]
Looks exactly like this test case, only it seems to work there: https://github.com/django-waffle/django-waffle/blob/3c5e93fa360c250af953a9abb30c2141aa86bd15/waffle/tests/test_testutils.py#L39-L48
:thinking:

**Comment by nanorepublica** [comment]
@mhemeryck see #328 and #329. I assume you are using Django TestCase and not TransactionTestCase?

**Comment by mhemeryck** [comment]
If I understand correctly, this has something to do with some test cache (?) not being flushed for specific django test cases.

We use `pytest-django` for test cases. I don't have that much insight into how this library works from the inside, but looking at the docs (https://pytest-django.readthedocs.io/en/latest/helpers.html#pytest-mark-django-db-request-database-access) it seems some kind of wrapper indeed of `TestCase` rather than `TransactionTestCase`.

Tried it with this parameter on v0.15.1

```python
from pytest import mark
from waffle import switch_is_active
from waffle.testutils import override_switch

@mark.django_db(transaction=True)
def test_feature_flag_override_true():
    with override_switch("myswitch", active=True):
        assert switch_is_active("myswitch")


@mark.django_db(transaction=True)
def test_feature_flag_override_false():
    with override_switch("myswitch", active=False):
        assert not switch_is_active("myswitch")
```
... and indeed, both test cases pass now. @nanorepublica I would guess your fix for the "regular" test case would then also solve my issue.

**Comment by jaywhy13** [comment]
I was trying to understand why using `TransactionTestCase` makes the second test pass. I have a similar problem I was able to "fix" by using `transaction=True`. Without `transaction = True` pytest uses `TestCase` which rolls back the transaction after each test, a little less aggressive than `TransactionTestCase` which truncates all tables. Does the cache perform differently under `TranactionTestCase` somehow?

**Comment by tkremmel** [comment]
We stumbled over the same problem using `pytest-django`, running in a multi-core CI test setup. However we were not able to solve the problem of randomly failing test cases / switches having the wrong value using `@mark.django_db(transaction=True)`.

**Comment by mhemeryck** [comment]
tbh, we are currently still on v0.14.0 since I didn't want to change all test cases to use the `TransactionTestCase` where it wasn't required before. I've seen that there have been a number of releases since then, but I'm not sure whether this has already been fixed?

**Update**: just checked (part) of our test suite against the latest version (v0.17) and it seems to be fixed here (might have been fixed in some earlier version even).

Will have a look later on whether we can do the update, since I've also noticed some new auto-create behavior was added (not sure whether we need / want this).

Anyways, the original issue seems to be fixed for me, so unless anyone objects, I would like to close this issue for you, OK?
